media/js/pocket/updates-signup.es6.js

```diff
@@ -0,0 +1,120 @@
+        } else {
+            // backend error, tampered form/missing fields
+            error = form.querySelector('.error-try-again-later');
+        }
+        if (error) {
+            error.classList.remove('hidden');
+        }
+    },
+
+    handleFormSuccess: () => {
+        const formWrapper = document.getElementById(
+            'pocket-updates-form-wrapper'
+        );
+        const thanksWrapper = document.getElementById('pocket-updates-thanks');
+        formWrapper.classList.add('hidden');
+        thanksWrapper.classList.remove('hidden');
+    },
+    postToBackend: (payload, url, successCallback, errorCallback) => {
+        const xhr = new XMLHttpRequest();
+
+        xhr.onload = function (e) {
+            let response = e.target.response || e.target.responseText;
+
+            if (typeof response !== 'object') {
+                response = JSON.parse(response);
+            }
+
+            if (response) {
+                if (
+                    response.status === 'success' &&
+                    e.target.status >= 200 &&
+                    e.target.status < 300
+                ) {
+                    successCallback();
+                } else {
+                    errorCallback(response);
+                }
+            } else {
+                errorCallback();
+            }
+        };
+        xhr.onerror = errorCallback;
+        xhr.open('POST', url, true);
+        xhr.setRequestHeader('Content-type', 'application/json');
+        xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
+        xhr.timeout = 5000;
+        xhr.ontimeout = errorCallback;
+        xhr.responseType = 'json';
+        xhr.send(JSON.stringify(payload));
+    },
+
+    subscribe: (e) => {
+        const payload = Object.fromEntries(new FormData(form).entries());
+        const url = form.getAttribute('action');
+
+        e.preventDefault();
+        e.stopPropagation();
+
+        // Disable form fields until POST has completed.
+        FormUtils.disableFormFields(form);
+
+        // Clear any prior messages that might have been displayed.
+        FormUtils.clearFormErrors(form);
+
+        // Submit to the server
+        UpdatesForm.postToBackend(
+            payload,
+            url,
+            UpdatesForm.handleFormSuccess,
+            UpdatesForm.handleFormError
+        );
+    },
+
+    getSearchParams: () => {
+        // isolated to make easier to mock in tests
+        return new URL(window.location.href).searchParams;
+    },
+
+    populateFromQuerystring: () => {
+        const soughtParams = ['campaign', 'medium', 'source'];
+        // Grab the UTM params for the hidden form fields, if we can, set them
+        // as default form values
+        for (const [key, value] of UpdatesForm.getSearchParams()) {
+            const strippedKey = key.replace('utm_', '');
+            if (soughtParams.includes(strippedKey)) {
+                form.elements[strippedKey].value = value;
```

Comment:
Question: do we need to do any checks here to avoid possible XSS? It might be worth adding a regex that only allows alpha numeric characters up to a certain length.

Maybe something like `/^[\w/.%-]{1,100}$/`?

Reply:
@alexgibson I've gone with encodeURIComponent to render the string inert, rather than trust I can write an all-encompassing XSS regex 😆 -- what do you think [of it](https://github.com/mozilla/bedrock/pull/13154/commits/bb751212f697c1f5e4888e94930b6d351527c5b2)?

Reply:
@stevejalim it definitely helps (and we should definitely encode the values either way), although the above regex is also pretty simple / straight forward. Is there something you're unsure about there?

Reply:
We could also avoid injecting the values into the DOM right? We can just hold these values in a variable in the script.

Reply:
@alexgibson regexery added in 88394d3611216c9b7e9e67da3519639be035becc -- including a couple more characters if you think they're acceptable (underscores and integers might appear in campaign codes, I'm assuming) 🤞

Reply:
@stevejalim the original regex i added above should capture alpha numeric characters (including underscores) I think

Reply:
@alexgibson Whoops 🤦 - fixing now. Cheers!